### ISSUE TITLE
Import notifications

### ISSUE THREAD

**Issue by lcardito** [rollbar/terraform-provider-rollbar#240]
This relates to https://github.com/rollbar/terraform-provider-rollbar/pull/236

I am trying to import an already existing resource and for some reason the provider can't read it from the rollbar API. I've poked a little around and I can't exactly see what's going wrong.

Some data :
```
❯ terraform --version
Terraform v1.0.6
on darwin_amd64
+ provider registry.terraform.io/rollbar/rollbar v1.2.0
```

My resource configuration is the following:

```
resource "rollbar_notification" "pd_dev_new_incident" {
  channel = "pagerduty"
}
```

Then I am trying to import using the notification id:

```
TF_LOG=TRACE terraform import rollbar_notification.pd_dev_new_incident "5141495"
```

The interesting provider logs I think are:

```
2021-09-13T12:29:50.176+0200 [DEBUG] provider.terraform-provider-rollbar_v1.2.0: {"level":"info","id":5141495,"time":"2021-09-13T12:29:50+02:00","message":"Reading rollbar_notification resource"}
2021-09-13T12:29:50.176+0200 [DEBUG] provider.terraform-provider-rollbar_v1.2.0: {"level":"debug","notificationID":5141495,"time":"2021-09-13T12:29:50+02:00","message":"Reading notification from API"}
2021-09-13T12:29:50.519+0200 [DEBUG] provider.terraform-provider-rollbar_v1.2.0: {"level":"error","notificationID":5141495,"error":"not found","time":"2021-09-13T12:29:50+02:00"}
2021-09-13T12:29:50.519+0200 [DEBUG] provider.terraform-provider-rollbar_v1.2.0: {"level":"info","id":5141495,"time":"2021-09-13T12:29:50+02:00","message":"Notification not found - removed from state"}
rollbar.rollbar_notification.pd_dev_new_incident result": visit complete
```

I ran a quick test with then API and the notification seems to be there..

```
❯ curl -s "https://api.rollbar.com/api/1/notifications/pagerduty/rule/5141495" \
    -H "Content-Type: application/json" \
    -H "X-Rollbar-Access-Token: XXXXXXXXXXXXXXXXXX" \
    | jq '.result.id'

5141495
```

**Comment by lcardito** [comment]
I've done a little digging (first time I get into the implementation of a terraform provider) and I was probably a little too optimistic in expecting the `import` to work out of the box. This might probably be a different feature

**Comment by lcardito** [comment]
@pawelsz-rb mentioning you here so we can continue our conversation.

The bottom line is that the `ImportStatePassthroughContext` used in the `ResourceImporter` https://github.com/rollbar/terraform-provider-rollbar/blob/master/rollbar/resource_notification.go#L47-L49 is only passing down the `id` of the resource specified in the `import` command. This means that the code can't build the right URL ending up always in a 404 😢

**Comment by awendt** [comment]
If I look at the (slightly complex) [`security_group_rule` import](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/security_group_rule#import) in the AWS provider, I can see it's using a custom format for importing because this also needs more than just the ID.

It seems it's up to the provider to turn the `ID` param in `terraform import ADDRESS ID` into something usable…

Its `ResourceImporter` is used like this: https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/resource_aws_security_group_rule.go#L31-L42

**Comment by lcardito** [comment]
Yeah - I was thinking this project should use a `channel:notificationId` type of `ID` for the `import` command. Not sure tho if it has some other side-effect in the terraform provider overall.

**Comment by awendt** [comment]
I had a look at how Github implements `terraform import github_membership` and realized their ID for importing and in the state file _are the same thing_. This way, `resourceGithubMembershipRead` can always assume it's a two-part ID.

I assume #236 writes just the ID into the state file (I would have to double-check but that's going to take while). If so, fixing it would be a breaking change.

**Comment by pawelsz-rb** [comment]
> If I look at the (slightly complex) [`security_group_rule` import](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/security_group_rule#import) in the AWS provider, I can see it's using a custom format for importing because this also needs more than just the ID.
> 
> It seems it's up to the provider to turn the `ID` param in `terraform import ADDRESS ID` into something usable…
> 
> Its `ResourceImporter` is used like this: https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/resource_aws_security_group_rule.go#L31-L42

**Comment by pawelsz-rb** [comment]
> If I look at the (slightly complex) [`security_group_rule` import](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/security_group_rule#import) in the AWS provider, I can see it's using a custom format for importing because this also needs more than just the ID.
> 
> It seems it's up to the provider to turn the `ID` param in `terraform import ADDRESS ID` into something usable…
> 
> Its `ResourceImporter` is used like this: https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/resource_aws_security_group_rule.go#L31-L42

**Comment by pawelsz-rb** [comment]
> @pawelsz-rb mentioning you here so we can continue our conversation.
> 
> The bottom line is that the `ImportStatePassthroughContext` used in the `ResourceImporter` https://github.com/rollbar/terraform-provider-rollbar/blob/master/rollbar/resource_notification.go#L47-L49 is only passing down the `id` of the resource specified in the `import` command. This means that the code can't build the right URL ending up always in a 404 😢

all the resources are built with ImportStatePassthroughContext, so I kept it that way for consistency

**Comment by pawelsz-rb** [comment]
> I had a look at [how Github implements `terraform import github_membership`](https://github.com/integrations/terraform-provider-github/blob/d80ff814cdce916ebf05d47210a25c1e963542fb/github/resource_github_membership.go) and realized their ID for importing and in the state file _are the same thing_. This way, `resourceGithubMembershipRead` can always assume it's a two-part ID.
> 
> I assume #236 writes just the ID into the state file (I would have to double-check but that's going to take while). If so, fixing it would be a breaking change.

notification resource writes ID to the state file, just like any other resource. Notification resource was just added on the top of other resources, and implemented in the same way. if you see some differences in implementation between our existing resources and this resource then please let me know.

**Comment by awendt** [comment]
> notification resource writes ID to the state file, just like any other resource. Notification resource was just added on the top of other resources, and implemented in the same way. if you see some differences in implementation between our existing resources and this resource then please let me know.

My understanding is this:

1. I don't doubt that the resource is implemented in the same way as other resources
2. I'm wondering if this the right approach here, given that `terraform import` needs a tuple (channel, ID) to work correctly
3. Other providers are using two-part IDs for similar functionality
4. I think a better ID for the resource would be `channel:notificationId`
5. I'm having a hard time seeing how this can be achieved, given that a numerical ID is already released for notifications.

@pawelsz-rb Do you see any way to make `terraform import` work with notifications, without changing the ID format for this particular resource?

Don't get me wrong: I'm thrilled to see that we can manage notifications using Terraform, I'm less thrilled about the prospect of manually deleting and subsequently terraforming hundreds of notifications across many Rollbar projects (because I can't import them). This is my motivation to discuss how we can get the import to work.

**Comment by pawelsz-rb** [comment]
> > notification resource writes ID to the state file, just like any other resource. Notification resource was just added on the top of other resources, and implemented in the same way. if you see some differences in implementation between our existing resources and this resource then please let me know.
> 
> My understanding is this:
> 
> 1. I don't doubt that the resource is implemented in the same way as other resources
> 2. I'm wondering if this the right approach here, given that `terraform import` needs a tuple (channel, ID) to work correctly
> 3. Other providers are using two-part IDs for similar functionality
> 4. I think a better ID for the resource would be `channel:notificationId`
> 5. I'm having a hard time seeing how this can be achieved, given that a numerical ID is already released for notifications.
> 
> @pawelsz-rb Do you see any way to make `terraform import` work with notifications, without changing the ID format for this particular resource?
> 
> Don't get me wrong: I'm thrilled to see that we can manage notifications using Terraform, I'm less thrilled about the prospect of manually deleting and subsequently terraforming hundreds of notifications across many Rollbar projects (because I can't import them). This is my motivation to discuss how we can get the import to work.

@awendt , this is already implemented, and it takes channel,notificationId, instead or channel:notificationId, I went through terraform documentation and they allow prefix per their structure. So the newest release 1.4.0 should take care of that. Please let me know if you have any other questions

**Comment by pawelsz-rb** [comment]
closing issue - resolved
